Wrapper status is not updated correctly

VindeeR opened the pull request:
Closes #2274.

File generation that AS will look for to try to update the status is created with a different name than what it searches for, this was creating a conflict on the validation and not updating it properly.

The file name is being reverted to how it used to be, and if needed the change it can be done by changing all the locations that it will impact

dbeltrankyl commented on autosubmit/job/job_common.py:
```diff
@@ -124,16 +126,16 @@ def as_header(scheduler_header, executable):
             set -xuve
             job_name_ptrn='%CURRENT_LOGDIR%/%JOBNAME%'
-            echo $(date +%s) > ${job_name_ptrn}_STAT_%FAIL_COUNT%
+            echo $(date +%s) > ${job_name_ptrn}_STAT
```
I think this change will mess up up the log recovery and timestamps

dbeltrankyl commented on autosubmit/job/job_common.py:
```diff
@@ -154,7 +156,7 @@ def as_tailer():
                 ###################
                 set -xuve
-                echo $(date +%s) >> ${job_name_ptrn}_STAT_%FAIL_COUNT%
+                echo $(date +%s) >> ${job_name_ptrn}_STAT
```
I think this change will mess up the log recovery and timestamps

dbeltrankyl commented on autosubmit/job/job_common.py:
```diff
@@ -166,12 +168,13 @@ class StatisticsSnippetPython:
+    def as_header(self, scheduler_header, executable):
         if not executable:
-            executable = "/usr/bin/env python"+str(self.version)
+            executable = "/usr/bin/env python" + str(self.version)
```
Since you did some lint changes, maybe you can also put this as an f-string

dbeltrankyl commented on autosubmit/job/job_common.py:
```diff
@@ -367,17 +370,17 @@ def parse_output_number(string_number):
     """
     Receives the wallclock times an increases it according to a quantity times the number of the current chunk.
-    The result cannot be larger than 48:00. 
+    The result cannot be larger than 48:00.
```
This 48:00, should be: cannot be large and the platform max_wallclock

dbeltrankyl reviewed (changes requested):
Hello @VindeeR 

Thanks for looking at it!

I didn't know this would mess with the status. 

However, I would do the reverse of what you propose. Instead of changing the STAT name that the cmd writes, change the function that checks it.  

The `%FAIL_COUNT%` was added because otherwise, we will overwrite the STAT FILES if a job is FAILED -> RETRIED -> FAILED/COMPLETED. 

A proper solution would be to change the function that looks for this stat file to consider the job's fail count. 

You can obtain simple jobs and non-vertical wrapper jobs with only the job.fail_count.

Vertical-wrapped jobs are trickier, and you may need to look at job.retrials.

The retrials for the vertical jobs are being made inside the same wrapper without resubmitting each job manually, so Autosubmit can't update the fail_count value, as it does not manage it. 

However, maybe it is enough if you look at the first retrial ( not sure now if it starts with _0 or _%job.retrials% in that case)

dbeltrankyl commented on autosubmit/platforms/platform.py:
```diff
@@ -718,16 +718,16 @@ def get_stat_file(self, job, count=-1):
             filename = job.stat_file + "0"
         else:
-            filename = job.name + '_STAT_{0}'.format(str(count))
+            filename = f'{job.name}_STAT_{job.fail_count}'
```
Hello @VindeeR 

Is that where the issue was?

I find it weird, because it should still look for the count that is passed by the function.


It is called by: 
```python
        # Write stats for vertical wrappers
        if self.wrapper_type == "vertical":  # Disable AS retrials for vertical wrappers to use internal ones
            for i in range(0, int(last_retrial + 1)):
                self.platform.get_stat_file(self, count=i)
                self.write_vertical_time(i)
                self.inc_fail_count()
```

The job.fail_count in that case is always 0 

>     

```python
if count == -1:  # No internal retrials
    filename = job.stat_file + "0"
```

I think the change should be there; instead of "0" it should be job.fail_count

dbeltrankyl replied:
The change I mention won't solve the issue with the monitor tho, ( it is a bug with the timestamps  for the other jobs seems )

I'm not sure how your change solves the issue with the monitor :thinking: 

As the code that calls it, should already gather the stat 0, but also the rest of the retries

dbeltrankyl replied:
Also, this is inside the log process recovery, which means that will be only called once the job is finished ( FAILED/COMPLETED)

VindeeR replied:
No this is not where the issue was, this is just me updating to the f-string format to make it standard.

dbeltrankyl commented:
Hello @VindeeR 

I think the issue may be here: 

```
    def _check_running_jobs(self): # <- this function
           ....
            # PREPARE SCRIPT TO SEND
            # When a inner_job is running? When the job has an _STAT file
            command = textwrap.dedent("""
            cd {1}
            for job in {0}
            do
                if [ -f "${{job}}_STAT" ] # <- 
                then
                        echo ${{job}} $(head ${{job}}_STAT) # <-
                else
                        echo ${{job}}
                fi
            done
            """).format(str(not_finished_jobs_names), str(remote_log_dir), '\n'.ljust(13))
```

dbeltrankyl commented:
About the proposed change, it is not enough for vertical wrappers. 

It will work well as long as there are no failed retries. 

Once there are failed retries, the job will be marked as FAILED because that script will check only the first stat file and see that the job is finished.

So you should also modify the script to look for:

When job.retrials > 0...
1) Look for more STAT files for that inner_job, and check the last one.
2) If there are no more stat files, you must ensure they are not because the next retry has not started. 

2.a) One idea is to look for the WRAPPER_FAILED file. If it exists, it is because the wrapper has ended 
2.b) Another idea is just to mark it as running until the next job in the vertical list is != QUEUING

dbeltrankyl commented:
After this, we will need to implement some tests. It will be hard to test horizontal, horizontal-vertical, and vertical-horizontal, so maybe you could just test them manually. 

For the vertical ones, it would be a matter of improving or adding a new integration test to monitor the status of the jobs. You can do it by:

* Launch the autosubmit run in the background or run it for a while and stop it 

Check this file (jobs_active_status.log)..

@laptop:~/autosubmit/a00x/tmp/ASLOGS$ cat jobs_active_status.log*
## String representation of Job List [1]  ## 
a00x_SIM [READY] 

Or you can load the job_list and pool the status from there

The integration tests generate a real experiment stored in the tmp_dir folder. They don't mock anything.

VindeeR commented:
> Hello @VindeeR
> 
> I think the issue may be here:
> 
> ```python
>     def _check_running_jobs(self): # <- this function
>            ....
>             # PREPARE SCRIPT TO SEND
>             # When a inner_job is running? When the job has an _STAT file
>             command = textwrap.dedent("""
>             cd {1}
>             for job in {0}
>             do
>                 if [ -f "${{job}}_STAT" ] # <- add the fail_count
>                 then
>                         echo ${{job}} $(head ${{job}}_STAT) # <- add the fail_count
>                 else
>                         echo ${{job}}
>                 fi
>             done
>             """).format(str(not_finished_jobs_names), str(remote_log_dir), '\n'.ljust(13))
> ```

Yes, the issue was there, but yesterday you told me to, rolling it back to how it was trying to make the other part of the code look for the new file it creates (_STAT_0 | _STAT_1), so I reverted this line and changed the other part of the code to look for the file

dbeltrankyl commented:
Hello @VindeeR 

Can you review this checklist and mark what you've tested or what still needs to be tested? 

What I do with the wrappers when I mess with retries ( that's another reason why it is hard to test)

Is running autosubmit run with: 


* [ ] horizontal | 1 retries | "COMPLETED" # doesn't matter the tries it should only do one
* [ ] horizontal-vertical | 1 retries | "COMPLETED" # doesn't matter the tries it should only do one
* [ ] vertical-horizontal | 1 retries | "COMPLETED" # doesn't matter the tries it should only do one
* [ ] vertical | 1 retries | "COMPLETED" # doesn't matter the tries it should only do one

* [ ] horizontal | 0 retries | "FAILED"
* [ ] horizontal | 1 retries | "FAILED"  # fail all 
* [ ] horizontal | 1 retries | "FAILED->COMPLETED # fail first retry

This three times.. one with the fail point at chunk 0, another with the fail point at chunk 1, and another with the fail point at chunk 2:

First chunk ( 1)

* [ ] horizontal-vertical | 1 retries | "FAILED"  # fail all 
* [ ] vertical-horizontal | 1 retries | "FAILED"  # fail all 
* [ ] vertical | 1  retries | "FAILED"  # fail all 

* [ ] horizontal-vertical | 1 retries | "FAILED->COMPLETED # fail the first time
* [ ] vertical-horizontal | 1 retries | "FAILED->COMPLETED # fail the first time
* [x] vertical | 1  retries | "FAILED->COMPLETED" * We can't test this case, I think, because the fail_count will always be 0 

Second chunk ( 2)

* [ ] horizontal-vertical | 1 retries | "FAILED"  # fail all 
* [ ] vertical-horizontal | 1 retries | "FAILED"  # fail all 
* [ ] vertical | 1  retries | "FAILED"  # fail all  

* [ ] horizontal-vertical | 1 retries | "FAILED->COMPLETED # fail the first time
* [ ] vertical-horizontal | 1 retries | "FAILED->COMPLETED # fail the first time 
* [x] vertical | 1  retries | "FAILED->COMPLETED" * We can't test this case, I think, because the fail_count will always be 0 

Last chunk ( 3)

* [ ] horizontal-vertical | 1 retries | "FAILED"  # fail all 
* [ ] vertical-horizontal | 1 retries | "FAILED"  # fail all 
* [ ] vertical | 1  retries | "FAILED"  # fail all 

* [ ] horizontal-vertical | 1 retries | "FAILED->COMPLETED # fail the first time
* [ ] vertical-horizontal | 1 retries | "FAILED->COMPLETED # fail the first time
* [x] vertical | 1  retries | "FAILED->COMPLETED" * We can't test this case, I think, because the fail_count will always be 0 


How to do it? 

```
EXPERIMENT:
  DATELIST: 19880101
  MEMBERS: fc0 fc1
  CHUNKSIZEUNIT: month
  CHUNKSIZE: 1
  NUMCHUNKS: 3
  CALENDAR: standard
  SPLITSIZEUNIT: day
  SPLITSIZE: 1
  SPLITPOLICY: flexible


JOBS:
  SIMV:
    DEPENDENCIES:
      SIMV-1: {}
    SCRIPT: | 
      sleep 1 # to see anything
      if [ %FAIL_COUNT% -eq 0 ] && [ %CHUNK% -eq 1 ]; then  # change the number to select the chunk
        decho "forcefail"
      else
        echo "valid"
      fi
    WALLCLOCK: "00:10"
    RUNNING: chunk
    CHECK: on_submission
    RETRIALS: 1

  SIMH:
    SCRIPT: |
      sleep 1 # to see anything
      if [ %FAIL_COUNT% -eq 0 ] && [ %CHUNK% -eq 1 ]; then  # change the number to select the chunk
        decho "forcefail"
      else
        echo "valid"
      fi
    WALLCLOCK: "00:10"
    RUNNING: chunk
    CHECK: on_submission
    RETRIALS: 1

  SIMHV:
    SCRIPT: |
      sleep 1 # to see anything
      if [ %FAIL_COUNT% -eq 0 ] && [ %CHUNK% -eq 1 ]; then  # change the number to select the chunk
        decho "forcefail"
      else
        echo "valid"
      fi
    DEPENDENCIES:
      SIMHV-1: { }
    WALLCLOCK: "00:10"
    RUNNING: chunk
    CHECK: on_submission
    RETRIALS: 1
  SIMVH:
    SCRIPT: |
      sleep 1 # to see anything
      if [ %FAIL_COUNT% -eq 0 ] && [ %CHUNK% -eq 1 ]; then  # change the number to select the chunk
        decho "forcefail"
      else
        echo "valid"
      fi
    DEPENDENCIES:
      SIMVH-1: { }
    WALLCLOCK: "00:10"
    RUNNING: chunk
    CHECK: on_submission
    RETRIALS: 1


WRAPPERS:
  WRAPPERV:
    TYPE: vertical
    JOBS_IN_WRAPPER: SIMV
  WRAPPERHV:
    TYPE: horizontal-vertical
    JOBS_IN_WRAPPER: SIMHV
  WRAPPERVH:
    TYPE: vertical-horizontal
    JOBS_IN_WRAPPER: SIMVH
  WRAPPERH:
    TYPE: horizontal
    JOBS_IN_WRAPPER: SIMH
```


Note: I tested ( some cases ) in master branch and it complains about some logs not being recovered, although it does, there is an issue for it somewhere

dbeltrankyl commented:
( I left the sleep 1 by mistake, put a higher value so you can monitor it and check what happens )

kinow reviewed (commented):
I saw some changes I wrote recently in this PR and thought it weird. Looking at the list of commits & files modified, I think you may have had an issue rebasing/fixing conflicts...

VindeeR commented:
> I saw some changes I wrote recently in this PR and thought it weird. Looking at the list of commits & files modified, I think you may have had an issue rebasing/fixing conflicts...

I needed one commit from master on this branch so I did a rebase but did manage to cherry pick properly the commit, so I'll do a drop of the last 3 commits, thanks for warning @kinow <3

kinow commented on autosubmit/platforms/slurmplatform.py:
```diff
@@ -563,7 +563,8 @@ def get_submit_cmd(self, job_script: str, job, hold: bool=False, export: str="")
                 pass
-        return export + self.get_submit_cmd_x11(job.x11_options.strip(""), job_script.strip(""))
+        else:
+            return export + self.get_submit_cmd_x11(job.x11_options.strip(""), job_script.strip(""))
```
This function is now weird.

It's supposed to return a string, but it only returns a string if this `else` block is evaluated to `True`? In the other cases Python will make it return a `None`, I think. Is that correct?

dbeltrankyl replied:
That is okay. It is the revert I did #2325.

kinow commented on autosubmit/platforms/platform.py:
```diff
@@ -718,16 +718,16 @@ def get_stat_file(self, job, count=-1):
                 return True
-        Log.warning('{0}_STAT file not found', job.name)
+        Log.warning(f'{job.name}_STAT_{str(count)} file not found')
         return False
```
I think coverage report is complaining about this file as the number of uncovered lines is the same as before, but since we are changing this function, I think we could write a test for `get_stat_file`? Doesn't look too complicated, probably won't even require any mocking.

VindeeR replied:
new test finished could you please check it? @kinow

kinow reviewed (commented):
Only found in place in the code that the type returned seems to have changed, and left a comment about a function that could be tested. Otherwise, I think if @dbeltrankyl is OK with the changes for fail_count/count (he knows that part of the code a lot better), then it should be ready to be merged. Thanks!

dbeltrankyl commented:
Hello

What is the status of this? Is something pending from my side?

kinow commented:
> Hello
> 
> What is the status of this? Is something pending from my side?

Can you review the part about the wrappers, @dbeltrankyl ? I had a look last week because I was touching the same file in another branch. I'll have a look at what @VindeeR changed since my last comment, but it was mainly about tests :+1:

kinow commented on test/unit/test_platform.py:
```diff
@@ -0,0 +1,65 @@
+import pytest
+from autosubmit.platforms.locplatform import LocalPlatform
+
+@pytest.mark.parametrize(
```
Two blank lines (see wiggly lines in PyCharm)

VindeeR replied:
is there 2 blank lines to you? that´s weird it should have only 1, there´s only 1 to me

kinow commented on test/unit/test_platform.py:
```diff
@@ -0,0 +1,65 @@
+from pathlib import Path
+from test.unit.test_job import TestJob, FakeBasicConfig
+import pytest
+from autosubmit.platforms.locplatform import LocalPlatform
```
As this is a new file, maybe we should organize the imports. You can use isort, or PyCharm's optimize imports. With that, I get

```python
from pathlib import Path

import pytest

from autosubmit.platforms.locplatform import LocalPlatform
from test.unit.test_job import TestJob, FakeBasicConfig
```

Which is how I normally read imports. stdlib, other libs, and then the current module.

VindeeR replied:
Now when I use pylint it´s giving me the following error
`C0411: standard import "test.unit.test_job.TestJob" should be placed before third party import "pytest" and first party import "autosubmit.platforms.locplatform.LocalPlatform"  (wrong-import-order)`
Complaining about the order of the import

kinow commented on test/unit/test_platform.py:
```diff
@@ -0,0 +1,65 @@
+        [True,-1],
+        [True,0],
+        [False,-1],
+        [False,0]
```
Missing space after comma (PyCharm warning about this, you are not getting these warnings too?)

VindeeR replied:
no, nor by using pylint, in reality I just had vulture giving me an error about an attribute that could be unused `LOCAL_TMP_DIR` which is used by the test inside the function, maybe we can check which are the configs you have to replicate at the lint file

kinow replied:
This one of spaces is included in most basic configurations, and has been around in Python since the beginning I think. `pep8`, now `pycodestyle` finds it too.

```bash
$ pip install pycodestyle
$ pycodestyle test/unit/test_platform.py 
test/unit/test_platform.py:25:1: E302 expected 2 blank lines, found 1
test/unit/test_platform.py:28:14: E231 missing whitespace after ','
test/unit/test_platform.py:29:14: E231 missing whitespace after ','
test/unit/test_platform.py:30:15: E231 missing whitespace after ','
test/unit/test_platform.py:31:15: E231 missing whitespace after ','
test/unit/test_platform.py:36:80: E501 line too long (93 > 79 characters)
test/unit/test_platform.py:37:80: E501 line too long (116 > 79 characters)
test/unit/test_platform.py:54:80: E501 line too long (91 > 79 characters)
test/unit/test_platform.py:58:80: E501 line too long (100 > 79 characters)
test/unit/test_platform.py:63:80: E501 line too long (84 > 79 characters)
test/unit/test_platform.py:64:80: E501 line too long (93 > 79 characters)
```

When you install PyCharm in a new laptop and configures any interpreter, PyCharm will automatically give you this warning, unless you disable it. When there are warnings, you have the option to turn off warnings, and if you do that, you won't see warnings like this again. Maybe you can try resetting the code inspection settings in PyCharm settings.

kinow replied:
(ignore those 79 characters rule, it's old from the time monitor resolution was limited)

kinow reviewed (approved):
Tests looking good! Leaving final approval for @dbeltrankyl who reviewed the rest of the code. Left some comments about the new file added.

In PyCharm, when you open a file you can see the errors and warnings in the file at the right top corner.

![image](https://github.com/user-attachments/assets/6e6cfb09-7a7d-4b34-b776-625c1b6e34ae)

Your file is new, but had already 5 warnings, that are easy to be fixed, and PyCharm also points what's wrong. Normally these match pep8, and will probably help when we turn on the linter.

![image](https://github.com/user-attachments/assets/58a11cac-784f-4b03-a560-3b4fa0e2c0c2)

You can go through these wiggly lines under text, hovering the mouse to see what's wrong, or Inspect the code to see the warnings, or open the Errors log, or run some linter compatible with pep8, etc.

Fixing that you should get the green mark in this new added file.

![image](https://github.com/user-attachments/assets/5cb8a8e8-1b20-4850-8441-9bfa12d84f8e)

dbeltrankyl commented:
> Can you review the part about the wrappers, @dbeltrankyl ? I had a look last week because I was touching the same file in another branch. I'll have a look at what @VindeeR changed since my last comment, but it was mainly about tests 👍

The current code will work for all wrappers ( at least it seems in the checklist) 

However, something could be missing in the vertical ones.... but not sure how to test it as I don't know a way of forcing a failure only in the first case if it is a vertical wrapper...

The problem is that the fail_count for vertical wrapper is not updated in the Autosubmit process, only in the wrapper script. I'm not sure now if the fail_count will be the max_retrial number or it will be 0 when it reachs that point. 

I'll revise it later and/or apply a patch if neecesary, after releasing AS 4.1.14

VindeeR commented:
@dbeltrankyl after making the fixes suggested by @kinow can I merge?

dbeltrankyl commented:
> @dbeltrankyl after making the fixes suggested by @kinow can I merge?

Not yet, I want to try the vertical stuff

dbeltrankyl commented:
Testing using

```yaml
JOBS:
  SIMV:
    DEPENDENCIES:
      SIMV-1: {}
    SCRIPT: | 
      sleep 15 # to see anything
      if [ $(( $(date +%S) % 2 )) -eq 0 ]; then  # fail on even seconds
        decho "forcefail"
      else
        echo "valid"
      fi
    WALLCLOCK: "00:10"
    RUNNING: chunk
    CHECK: on_submission
    RETRIALS: 1
```

Then I just ran it 3 or 4 times.

I confess that the failure point is random. I don't know other way of doing it, but I managed to see a case in which the first retry failed and the second one was completed ( according to the log recovered jobs) and for Autosubmit, it was always running. 

So I think we can merge it. 

Can you add the changelog first? @VindeeR

dbeltrankyl commented:
Also 

```
            command = textwrap.dedent(f"""
            cd {str(remote_log_dir)}
            for job in {str(not_finished_jobs_names)}
            do
                if [ -f "${{job}}_STAT_{self.fail_count}" ]
                then
                        echo ${{job}} $(head ${{job}}_STAT_{self.fail_count})
                else
                        echo ${{job}}
                fi
            done
            """)
```

In case of a Vertical wrapper, the fail_count is always 0. It doesn't affect the outcome tho

dbeltrankyl commented:
@VindeeR can you prioritize the merge of this? I want to include it in v4.1.14. Thanks!

VindeeR commented:
> Testing using
> 
> ```yaml
> JOBS:
>   SIMV:
>     DEPENDENCIES:
>       SIMV-1: {}
>     SCRIPT: | 
>       sleep 15 # to see anything
>       if [ $(( $(date +%S) % 2 )) -eq 0 ]; then  # fail on even seconds
>         decho "forcefail"
>       else
>         echo "valid"
>       fi
>     WALLCLOCK: "00:10"
>     RUNNING: chunk
>     CHECK: on_submission
>     RETRIALS: 1
> ```
> 
> Then I just ran it 3 or 4 times.
> 
> I confess that the failure point is random. I don't know other way of doing it, but I managed to see a case in which the first retry failed and the second one was completed ( according to the log recovered jobs) and for Autosubmit, it was always running.
> 
> So I think we can merge it.
> 
> Can you add the changelog first? @VindeeR
> 
> Thanks for the work!

about the changelog should I create a new one for the 4.1.14? and for all changes from now on we should add there too right?
